File: Difficulty: Medium/Frog Jump/frog-jump.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    int f(int ind, vector<int>& height, vector<int> &dp){
        if(ind == 0) return 0;
        
        if(dp[ind] != -1) return dp[ind];
        
        int left = f(ind - 1, height, dp) + abs(height[ind] - height[ind - 1]);
        int right = INT_MAX;
        if(ind > 1){
            right = f(ind - 2, height, dp) + abs(height[ind] - height[ind - 2]);
        }
        
        return dp[ind] = min(left, right);
    }
  
    int minCost(vector<int>& height) {
        int n = height.size();
        vector<int> dp(n+1, -1);
        return f(n-1, height, dp);
    }
};


//{ Driver Code Starts.

int main() {
    string str;
    getline(cin, str);
    int t = stoi(str);
    while (t--) {
        getline(cin, str);
        stringstream ss(str);
        vector<int> arr;
        int num;
        while (ss >> num) {
            arr.push_back(num);
        }
        Solution ob;
        cout << ob.minCost(arr) << endl;
        cout << "~" << endl;
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Frog Jump/frog-jump.py
#User function Template for python3
class Solution:
    def helper(self, i, dp, height):
        if i == 0:
            return 0
        
        if dp[i] != -1:
            return dp[i]
        
        left = abs(height[i] - height[i-1]) + self.helper(i-1, dp, height)
        right = 1e9
        if i > 1:
            right = abs(height[i] - height[i-2]) + self.helper(i-2, dp, height)
        
        dp[i] = min(left, right)
        
        return dp[i]
    
    def minCost(self, height):
        n = len(height)
        dp = [-1] * n
        return self.helper(n-1, dp, height)